test/e2e/common.go

```diff
@@ -153,21 +154,77 @@ func (c *Config) GetVariable(varName string) string {
 	return value
 }
 
+func isUndesiredState(currentState metal3api.ProvisioningState, undesiredStates []metal3api.ProvisioningState) bool {
+	if undesiredStates == nil {
+		return false
+	}
+
+	for _, state := range undesiredStates {
+		if (state == "" && currentState == "") || currentState == state {
+			return true
+		}
+	}
+	return false
+}
+
 type WaitForBmhInProvisioningStateInput struct {
-	Client client.Client
-	Bmh    metal3api.BareMetalHost
-	State  metal3api.ProvisioningState
+	Client          client.Client
+	Bmh             metal3api.BareMetalHost
+	State           metal3api.ProvisioningState
+	UndesiredStates []metal3api.ProvisioningState
 }
 
 func WaitForBmhInProvisioningState(ctx context.Context, input WaitForBmhInProvisioningStateInput, intervals ...interface{}) {
 	Eventually(func(g Gomega) {
 		bmh := metal3api.BareMetalHost{}
 		key := types.NamespacedName{Namespace: input.Bmh.Namespace, Name: input.Bmh.Name}
 		g.Expect(input.Client.Get(ctx, key, &bmh)).To(Succeed())
-		g.Expect(bmh.Status.Provisioning.State).To(Equal(input.State))
+
+		currentStatus := bmh.Status.Provisioning.State
+
+		// Check if the current state matches any of the undesired states
+		if isUndesiredState(currentStatus, input.UndesiredStates) {
+			g.Expect(fmt.Sprintf("Encountered undesired state: %s", currentStatus)).ToNot(Succeed())
+			return
```

Comment:
This is not checking temporary undesired states. Eventually will just keep going after the return unless you tell it to `StopTrying`.
```suggestion
			StopTrying(fmt.Sprintf("BMH is in an unexpected state: %s", currentStatus)).Now()
```